contrib/filters/http/squash/source/config.cc

```diff
@@ -1,12 +1,13 @@
-#include "source/extensions/filters/http/squash/config.h"
+#include "contrib/filters/http/squash/source/config.h"
```

Comment:
I thought this will be `contrib/squash/source/filters/http/config.h`? Per https://docs.google.com/document/d/1yl7GOZK1TDm_7vxQvt8UQEAu07UQFru1uEKXM6ZZg_g/edit#heading=h.zgixgx1c4t8u

Reply:
IMO it makes more sense to use the full extension path that matches the normal extensions tree. I can make the doc more clear on this.

Reply:
I meant we want contrib/<extension_name>/ as a unit, have source and test under it. In that way we can group extensions that have to be used together (for example, something like dynamic_forward_proxy should have both filter and cluster extension under `contrib/dynamic_forward_proxy`.) It also makes CODEOWNER clearer instead of mixing multiple extensions in same tree.

Reply:
I did end up grouping source/test under the same name, but I still think if we just do things by name it will get pretty chaotic. I have a strong preference for matching the same directory structure as source/extensions. You don't think it will get extremely messy if we don't have this structure?

Reply:
> I did end up grouping source/test under the same name, but I still think if we just do things by name it will get pretty chaotic. I have a strong preference for matching the same directory structure as source/extensions. You don't think it will get extremely messy if we don't have this structure?

No, that was the whole point I proposed `contrib/<extension_name>` and I feel strongly that under every `contrib/extension_name` it should look like a standalone repository. Then we can give the CODEOWNERS responsibility of managing everything under it, if no one is maintaining we can just delete the whole directory.  It also has the flexibility of having multiple extension under same name like I stated above. We can have a guidance that under `contrib/<extension_name>` should still follow same structure of the main tree.

Reply:
I think this falls apart for when contrib extensions share common code, which they will immediately I think when moving both postgres and myself. So then what is done with common code?

I'm not going to die on this hill as I would like to get this merged, but I don't think your proposal will work in all cases anyway, so I'm not sure it's worth changing to it.

Reply:
(Also, API is in another directory anyway.)

Reply:
(And also I just think it's confusing to have a top level directory of names with no organization. I think it will become unwieldy quickly.)

cc @envoyproxy/maintainers for further thoughts.

Reply:
With regards to API files, I think they should also be co-located with the source and tests for the extension, not in a separate `api` tree.

As far as structure, I think I agree with @lizan. Given the governing/social contract we're making for `contrib`, I think we should have a bunch of completely independent silo directories. For shared code, I think we could have another silo of `db_common` or something for shared postgres/mysql code (or similar); it's owners could be the union of the codeowners of each of those two extensions. Or each extension could have it's own entire copy of the shared code; that may end up being my preference.

Reply:
> With regards to API files, I think they should also be co-located with the source and tests for the extension, not in a separate api tree.

I tried to do this but it's not possible without a tremendous amount of work due to all of the API machinery. I'm sure it's possible but it's not work I have time to do, so if we want that someone else will need to pick it up. Note that I already cleared this with the API shepherds.

> Or each extension could have it's own entire copy of the shared code; that may end up being my preference.

This is a hill I'm willing to die on. I don't think code copying makes sense, so we must have some way of sharing code. For me personally I really don't understand how a giant top level mashed you tree is preferable to a nicely organized tree. Owners files can be used in either case, and we can track status in either case. So why not have nice organization?

As I said, I'm not going to demand the existing structure, though I strongly recommend it. I am going to demand some way of sharing code, so let me know what you all prefer there.

Reply:
I think the API stuff could be made to work colocated, but it's no big deal if we have it as independent. 

If structure is determined by extension structure, and there aren't any new extension types in `contrib/`, I think I'd be inclined to vote to mirror existing `source/extensions` structure. I think we should give folks in `contrib/` maximum freedom, but not where it can only lead to unnecessary disorganization.

Reply:
> As I said, I'm not going to demand the existing structure, though I strongly recommend it. I am going to demand some way of sharing code, so let me know what you all prefer there.

Regarding to sharing code, any sharing between contrib and core goes to core, so the question is how to share codes between contrib extensions. I'd vote to have another silos in `contrib/common/sqlutils`, this is no different from `source/extensions/common`. Looking at existing [`source/extensions/common`](https://github.com/envoyproxy/envoy/tree/main/source/extensions/common) probably only `sqlutils` will go contrib/common like that. For example even if we move redis to contrib (while we're not going to do), in the model that I proposed, the common code, network filter, health checker and cluster will be all under `contrib/redis`. The tree under `contrib/redis` can still match the existing structure though.

Reply:
> I'd vote to have another silos in contrib/common/sqlutils, this is no different from source/extensions/common

Yes, I agree, this is what I was going to do when I move that code.

> I proposed, the common code, network filter, health checker and cluster will be all under contrib/redis. The tree under contrib/redis can still match the existing structure though.

I understand the proposal, I just don't agree that it's worth making that directory incredibly messy. It's an extreme counterpoint, but imagine if all of the linux kernel drivers were in a single directory sorted by name of the driver only? That's what is being proposed here. I will try to ping all the maintainers in Slack so we can close this out. If the group consensus is to use this structure so be it.

Reply:
i think i tend to agree with @lizan altho separating every extension into its own directory means a lot of directory duplication in the envoy source, i think it probs makes more sense from the pov of extension builders to see examples that are entirely independent

i guess i get the point about linux kernel modules, but at the same time i think that is more of a comparison to core extensions rather than contrib ones

i guess there a couple of different motivations here - one is the maintainability/sanity of the envoy contrib source, and another is isolating extensions to make it clearer how the modularisation works to devs - if the latter is better served one way or other im inclined towards that way

if we do go down the path of separating each extension, there is also a question of extensions that extend other extensions and whether they then can/should be grouped

Reply:
I mildly prefer keeping code for a given extension in a given directory, though deleting the contrib/source/extensions/foo contrib/include/extensions/foo contrib/test/extensions/foo isn't a deal breaker (I'm fine either way)

I'm strongly against copy-pasted common code.  I'd lean towards a common directory with default visibility private, and explicit opt-in for each library so we can easily tell that common/foo_lib is used by extension A and B, and clean up if we delete extension A and B.

Reply:
Sorry I want to be completely clear about what is being proposed here since I'm not sure everyone is talking about the same thing.

The current code:

`/contrib/<... full hierarchical path based on extension type ...>/<extension name>/<source|test>/`

Lizan's proposal:

`/contrib/<extension name>/<... full hierarchical path based on extension type ...>//<source|test>/`

The latter means that when one looks in the contrib folder they will see a giant list of names with no context beyond that.

Reply:
The reason I suggested copy/paste shared code, is that if it isn't clear who the owner of the shared code is, what happens if one of the extensions using it wants to make a change and the other user doesn't? But I'm fine with the group-consensus to not duplicate the shared code.

Reply:
I think one way to looking at the `/contrib/extension-name/full` is like github organizations. The `extension-name` part is like a github org; then there's >=1 extensions underneath it, all owned by the same logical owner.

I think both proposals have merit; I'd be fine with either approach, and as we get more experience with managing this, and if we feel like it, we can re-org it into the other approach if needed. So do whatever is most expedient.